tests/Credentials/CredentialProviderTest.php

```diff
@@ -897,6 +897,108 @@ public function testEnsuresSourceProfileHasCredentials()
         }
     }
 
+
+    public function testSsoProfileProvider()
+    {
+        $dir = $this->clearEnv();
+        $ini = <<<EOT
+[default]
+sso_start_url = url.co.uk
+sso_region = us-west-2
+sso_account_id = 12345
+sso_role_name = roleName
+EOT;
+        $tokenFile = <<<EOT
+{"startUrl" : "url.com", "accessToken" : "token"}
+EOT;
+
+        $filename = $dir . '/config';
+        file_put_contents($filename, $ini);
+        $tokenFileName = $dir . "/sso/cache/" . sha1("url.co.uk") . '.json';
+        file_put_contents(
+            $tokenFileName, $tokenFile
+        );
+        putenv('HOME=' . dirname($dir));
+        $filename = $dir . '/config';
+        putenv('HOME=' . dirname($dir));
+
+        $result = [
+            'roleCredentials' => [
+                'accessKeyId'     => 'foo',
+                'secretAccessKey' => 'assumedSecret',
+                'sessionToken'    => null,
+                'expiration'      => DateTimeResult::fromEpoch(time() + 1000)
+            ],
+        ];
+
+        $sso = $this->getTestClient('Sso', ['credentials' => false]);
+        $this->addMockResults($sso, [
+            new Result($result)
+        ]);
+
+        try {
+            $creds = call_user_func(CredentialProvider::sso(
+                'default',
+                $filename,
+                ['ssoClient' => $sso])
+            )->wait();
+            $this->assertEquals('foo', $creds->getAccessKeyId());
+            $this->assertEquals('assumedSecret', $creds->getSecretKey());
+            $this->assertNull($creds->getSecurityToken());
+            $this->assertGreaterThan(
+                DateTimeResult::fromEpoch(time())->getTimestamp(),
+                $creds->getExpiration()->getTimestamp()
+            );
+        } catch (\Exception $e) {
+            throw $e;
+        } finally {
+            unlink($dir . '/config');
+            unlink($tokenFileName);
+        }
+    }
+
+    /**
+     * @expectedException \Aws\Exception\CredentialsException
+     * @expectedExceptionMessage  Cannot read credentials from
+     */
+    public function testSsoProfileProviderBadFile()
+    {
+        $dir = $this->clearEnv();
+
+        $filename = $dir . '/config';
+        putenv('HOME=' . dirname($dir));
+
+        try {
+            call_user_func(CredentialProvider::sso('default', $filename))->wait();
+        } catch (\Exception $e) {
+            throw $e;
+        }
+    }
+
+    /**
+     * @expectedException \Aws\Exception\CredentialsException
+     * @expectedExceptionMessage  must contain the following keys: sso_start_url, sso_region, sso_account_id, and sso_role_name
```

Comment:
Yes, that was intentional.  Since 'expectedExceptionMessage' just does a contains, and the message is variable with the filename as shown below.  This takes the most significant part of the message and checks for it.  The other option was using @expectedExceptionMessageRegExp which would add a bit of overhead that doesn't seem necessary.   It's just a small amount and only happens when tests are run, so I can add it if you'd prefer

                   ```
 "Profile {$ssoProfileName} in {$filename} must contain the following keys: "
                    . 'sso_start_url, sso_region, sso_account_id, and sso_role_name'
```